libnavigation-core/src/main/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetry.kt

```diff
@@ -0,0 +1,301 @@
+                }
+            }
+        }
+    }
+
+    private suspend fun getDatePartitionedLocations(predicate: (Location) -> Boolean): Array<Location> {
+        val retVal = mutableListOf<Location>()
+        channelLocationBuffer.send(LocationBufferControl(LocationBufferCommands.BUFFER_READ, null) { location ->
+            if (predicate(location)) {
+                retVal.add(location)
+            }
+        })
+        return retVal.toTypedArray()
+    }
+
+    private suspend fun parseRouteProgress(): TelemetryStep {
+        return TelemetryStep().apply {
+            val routeProgress = channelOnRouteProgress.receive().also { routeProgress ->
+                distanceRemaining = routeProgress.distanceRemaining().toInt()
+                durationRemaining = routeProgress.durationRemaining().toInt()
+            }
+
+            ifNonNull(routeProgress.currentLegProgress()?.upcomingStep()) { upcomingStep ->
+                distance = upcomingStep.distance().toInt()
+                duration = upcomingStep.duration().toInt()
+                upcomingType = upcomingStep.maneuver().type() ?: ""
+                upcomingModifier = upcomingStep.maneuver().modifier() ?: ""
+                upcomingName = upcomingStep.name() ?: ""
+            }
+
+            ifNonNull(routeProgress.currentLegProgress()?.currentStepProgress()?.step()) { currentStep ->
+                previousModifier = currentStep.maneuver().modifier() ?: ""
+                previousName = currentStep.name() ?: ""
+                previousType = currentStep.maneuver().type() ?: ""
+            }
+        }
+    }
+
+    /**
+     * TODO:OZ add one of these for each telemetry event
+     * Generates a well-formed Feedback event.
+     */
+    private suspend fun populateUserFeedbackEvent(event: TelemetryEventFeedback): MetricEvent =
+            TelemetryUserFeedbackWrapper(event,
+                    TelemetryUtils.retrieveVendorId(),
+                    NavigationUtils.obtainAudioType(context),
+                    getDatePartitionedLocations { location -> location.time < Time.SystemImpl.millis() },
+                    getDatePartitionedLocations { location -> location.time > Time.SystemImpl.millis() },
+                    UUID.randomUUID().toString(),
+                    event.screenShot,
+                    parseRouteProgress()
+            ).toMetricEvent()
+
+    /**
+     * TODO:OZ add code to handle all other event types. Once implemented, instead of throwing an exception the code
+     * should log an error
+     */
+    private suspend fun populateTelemetryEventWrapper(event: TelemetryEventInterface): MetricEvent? =
+            when (event) {
+                is TelemetryEventFeedback -> populateUserFeedbackEvent(event)
+                else -> throw(NavigationException("Unsupported telemetry event"))
+            }
+
+    private fun validateAccessToken(accessToken: String?) {
+        if (accessToken.isNullOrEmpty() ||
```

Comment:
you might check it [here](https://www.dcode.fr/boolean-truth-table) 

```suggestion
        if (accessToken.isNullOrEmpty() 
          || ( !accessToken.toLowerCase(Locale.US).startsWith("pk.")
               || !accessToken.toLowerCase(Locale.US).startsWith("sk."))
```

Reply:
👍

Reply:
That would be incorrect. a token can start with pk or sk. If we use an "or" than if either one is false, the entire if-statement fails. We want it to fail only if neither is correct. Neither is modeled with "and". Like this:
`if (accessToken.isNullOrEmpty()
 || ( !accessToken.toLowerCase(Locale.US).startsWith("pk.")  && !accessToken.toLowerCase(Locale.US).startsWith("sk."))
`

Reply:
yep, you're right, thanks!